src/opf/utils.jl

```diff
@@ -0,0 +1,301 @@
+using MathOptSymbolicAD
+using SparseArrays
+
+struct OPFData
+    case::String
+
+    N::Int  # number of buses
+    E::Int  # number of branches
+    G::Int  # number of generators
+    L::Int  # number of loads
+
+    Ag::SparseMatrixCSC{Float64,Int}  # generator incidence matrix
+
+    # Bus data
+    vmin::Vector{Float64}
+    vmax::Vector{Float64}
+    gs::Vector{Float64}  # shunt
+    bs::Vector{Float64}  # shunt
+    pd::Vector{Float64}  # nodal active power load
+    qd::Vector{Float64}  # nodal reactive power load
+    bus_arcs_fr::Vector{Vector{Int}}  # indices of branches exiting the bus
+    bus_arcs_to::Vector{Vector{Int}}  # indices of branches entering the bus
+    bus_gens::Vector{Vector{Int}}  # indices of generators at the bus
+    ref_bus::Int  # reference bus
+
+    # Generator data
+    pgmin::Vector{Float64}
+    pgmax::Vector{Float64}
+    qgmin::Vector{Float64}
+    qgmax::Vector{Float64}
+    c0::Vector{Float64}  # constant cost
+    c1::Vector{Float64}  # linear cost
+    c2::Vector{Float64}  # quadratic cost
+    gen_status::Vector{Bool}  # generator status
+
+    # Branch data
+    bus_fr::Vector{Int}  # from bus
+    bus_to::Vector{Int}  # to bus
+    g::Vector{Float64}
+    b::Vector{Float64}
+    gff::Vector{Float64}
+    gft::Vector{Float64}
+    gtf::Vector{Float64}
+    gtt::Vector{Float64}
+    bff::Vector{Float64}
+    bft::Vector{Float64}
+    btf::Vector{Float64}
+    btt::Vector{Float64}
+    smax::Vector{Float64}  # max apparent power flow
+    dvamin::Vector{Float64}  # angle difference min
+    dvamax::Vector{Float64}  # angle difference max
+    branch_status::Vector{Bool}  # branch status
+end
+
+"""
+    OPFData(network::Dict{String,Any})
+
+Convert a PowerModels data dictionary to `OPFData` structure.
+
+The PowerModels data dictionary must be in basic format.
+"""
+function OPFData(network::Dict{String,Any})
+    @assert network["per_unit"] == true "Network data is not per-unit scaled."
+    @assert network["baseMVA"] == 100.0 "Base MVA is not 100.0."
```

Comment:
Add a check that the network is in basic format, to avoid accidents if someone passes the raw files